### PR TITLE
[core] Use the ResourceTransform instead of callbacks

### PR DESCRIPTION
**Pull request by tmpsantos**
Based on the actor model, can be considered safer to use since we don't
need to worry about object lifetime anymore.

### CODE REVIEW

File: include/mbgl/actor/actor_ref.hpp [hunk 1]
```diff
@@ -24,19 +24,19 @@ template <class Object>
...
     void invoke(Fn fn, Args&&... args) {
         if (auto mailbox = weakMailbox.lock()) {
-            mailbox->push(actor::makeMessage(object, fn, std::forward<Args>(args)...));
+            mailbox->push(actor::makeMessage(*object, fn, std::forward<Args>(args)...));
         }
     }
 
 private:
-    Object& object;
+    Object* object;
```
Comment (jfirebaugh) [inline include/mbgl/actor/actor_ref.hpp]: What's the reason for this change?

File: include/mbgl/util/work_task_impl.hpp [hunk 2]
```diff
@@ -91,19 +91,4 @@ auto packageArgumentsAndCallback(std::shared_ptr<std::atomic<bool>> flag,
 }
 } // namespace detail
 
-template <class Fn, class... Args>
-std::shared_ptr<WorkTask> WorkTask::makeWithCallback(Fn&& fn, Args&&... args) {
```
Comment (jfirebaugh) [inline include/mbgl/util/work_task_impl.hpp]: You can delete `packageArgumentsAndCallback` above too.

File: include/mbgl/actor/actor_ref.hpp (hunk 1 above)
Comment (tmpsantos) [inline include/mbgl/actor/actor_ref.hpp]: Using references on attributes implicitly deletes the default constructor, so I would not be able to do:

```
void OnlineFileSource::setResourceTransform(optional<ActorRef<ResourceTransform>>&& transform) {
    resourceTransform = std::move(transform);
}
```

Comment (tmpsantos) [inline include/mbgl/actor/actor_ref.hpp]: Using references on attributes implicitly deletes the default assignment operator, so I would not be able to do:

```
void OnlineFileSource::setResourceTransform(optional<ActorRef<ResourceTransform>>&& transform) {
    resourceTransform = std::move(transform);
}
```

File: platform/darwin/src/MGLOfflineStorage.mm [hunk 3]
```diff
@@ -74,7 +78,8 @@ - (void)unpauseFileSource:(__unused NSNotification *)notification {
 - (void)setDelegate:(id<MGLOfflineStorageDelegate>)newValue {
     _delegate = newValue;
     if ([self.delegate respondsToSelector:@selector(offlineStorage:URLForResourceOfKind:withURL:)]) {
-        _mbglFileSource->setResourceTransform([offlineStorage = self](auto kind_, std::string&& url_) -> std::string {
+        delete _mbglResourceTransform;
+        _mbglResourceTransform = new mbgl::Actor<mbgl::ResourceTransform>(*mbgl::util::RunLoop::Get(), [offlineStorage = self](auto kind_, const std::string&& url_) -> std::string {
```
Comment (tmpsantos) [inline platform/darwin/src/MGLOfflineStorage.mm]: @frederoni I will need your help with this one. For instance, I just followed the "code style" here, but I would totally use `std::unique_ptr` if I could, like it was done for Android.

File: include/mbgl/util/work_task_impl.hpp (hunk 2 above)
Comment (tmpsantos) [inline include/mbgl/util/work_task_impl.hpp]: done

**Comment by tmpsantos** [comment]
:eyes: @frederoni @1ec5, I need help on the `darwin` part of this PR, please feel free to change it the way you find appropriated.

File: platform/darwin/src/MGLOfflineStorage.mm (hunk 3 above)
Comment (1ec5) [inline platform/darwin/src/MGLOfflineStorage.mm]: I’m not sure, but maybe we can replace the `mbglResourceTransform` property with an instance variable [here](https://github.com/mapbox/mapbox-gl-native/blob/46e5b4a0e4e1f4be7958417d9dc87d3763e71fc4/platform/darwin/src/MGLOfflineStorage.mm#L43):

```objc
@implementation MGLOfflineStorage {
    std::unique_ptr<Actor<ResourceTransform>> _mbglResourceTransform;
}
```
